templates/pritunl-plugin-ldap.j2

```diff
@@ -0,0 +1,118 @@
+        except ldap.LDAPError as e:
+            if type(e.message) == dict and 'desc' in e.message:
+                logger.error(e.message['desc'])
+            if type(e.message) == dict and 'desc' in e.message:
+                logger.error(e)
+            else:
+                logger.error(e)
+            raise
+        return l
+
+    def _unbind(self):
+        if self._conn is not None:
+            self.conn().unbind_s()
+            self._conn = None
+
+    def search(self, base_dn, searchFilter, searchAttribute=None, callback=None):
+        try:
+            ldap_result_id = self.conn().search(
+                                        base_dn,
+                                        ldap.SCOPE_SUBTREE,
+                                        searchFilter,
+                                        searchAttribute)
+            result_set = []
+            while 1:
+                result_type, result_data = self.conn().result(ldap_result_id, 0)
+                if (result_data == []):
+                    break
+                else:
+                    if result_type == ldap.RES_SEARCH_ENTRY:
+                        result_set += self._handleResults(result_data, callback)
+            self._unbind()
+            return result_set
+        except ldap.LDAPError as e:
+            self._unbind()
+            logger.error(e)
+
+    def _handleResults(self, data, callback):
+        results = []
+
+        for dn, record in data:
+            if dn is not None:
+                if callback is not None:
+                    results = callback(dn, record)
+                else:
+                    results.append((dn, record))
+
+        return results
+
+os.environ["PRITUNL_API_URL"] = "{{ pritunl_api.url }}"
+os.environ["PRITUNL_API_TOKEN"] = "{{ pritunl_api.token }}"
+os.environ["PRITUNL_API_SECRET"] = "{{ pritunl_api.secret }}"
+
+PRITUNL_ORG = "{{ pritunl_api.organization }}"
+GROUP_PATTERN = "{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='group_pattern') | join }}"
+
+URI = "{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='uri') | join }}"
+BIND_DN = "{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='bind_dn') | join }}"
+PASSWORD = "{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='password') | join }}"
+BASE_DN = "{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='base_dn') | join }}"
+GROUPS_DN = "{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='groups_dn') | join }}"
+SEARCH_ATTRIBUTE = ['{{pritunl_plugins | selectattr("name","equalto","ldap") | map(attribute='search_attribute') | join }}']
+
+L = LDAPSearch(URI, BIND_DN, PASSWORD)
+
+# ldapsearch callback
+def filter_groups(dn, record):
+    return [item.split(',')[0].replace('cn=', '') for item in record['memberOf'] if GROUP_PATTERN in item]
+
+def sso_authenticate(sso_type, host_id, host_name, user_name, user_email,
```

Comment:
Is this an callback plugin too?

Reply:
This is a pritunl handler, it think I'd better add it as a comment.

Reply:
Yeap, an docstring will be great! :+1:

Reply:
done